lib/Dashboard/CalendarWidget.php

```diff
@@ -97,14 +113,86 @@ public function getUrl(): ?string {
 		return null;
 	}
 
+	/**
+	 * @inheritDoc
+	 */
+	public function getIconUrl(): string {
+		return $this->urlGenerator->getAbsoluteURL(
+			$this->urlGenerator->imagePath(Application::APP_ID, 'calendar-dark.svg')
+		);
+	}
+
 	/**
 	 * @inheritDoc
 	 */
 	public function load(): void {
-		\OCP\Util::addScript('calendar', 'calendar-dashboard');
+		Util::addScript(Application::APP_ID, 'calendar-dashboard');
+		Util::addStyle(Application::APP_ID, 'dashboard');
 
-		$this->initialStateService->provideLazyInitialState(Application::APP_ID, 'dashboard_data', function () {
+		$this->initialStateService->provideLazyInitialState('dashboard_data', function () {
 			return $this->dataService;
 		});
 	}
+
+	/**
+	 * @inheritDoc
+	 *
+	 * @param string|null $since Use any PHP DateTime allowed values to get future dates
+	 * @param int $limit  Max 14 items is the default
+	 */
+	public function getItems(string $userId, ?string $since = null, int $limit = 14): array {
+		$calendars = $this->calendarManager->getCalendarsForPrincipal('principals/users/' . $userId);
+		$count = count($calendars);
+		if ($count === 0) {
+			return [];
+		}
+		$limitPerCalendar = (int)round(($limit / $count)) ?? 1;
+		$dateTime = (new DateTimeImmutable())->setTimestamp($this->timeFactory->getTime());
+		if ($since !== null) {
+			try {
+				$dateTime = new DateTimeImmutable($since);
+			} catch (\Exception $e) {
+				// silently drop the exception and proceed with "now"
+			}
+		}
+		$inTwoWeeks = $dateTime->add(new DateInterval('P14D'));
+		$options = [
+			'timerange' => [
+				'start' => $dateTime,
+				'end' => $inTwoWeeks,
+			]
+		];
+		$widgetItems = [];
+		foreach ($calendars as $calendar) {
+			$searchResult = $calendar->search('', [], $options, $limitPerCalendar);
+			foreach ($searchResult as $calendarEvent) {
+				/** @var DateTimeImmutable $startDate */
+				$startDate = $calendarEvent['objects'][0]['DTSTART'][0];
+				$widget = new WidgetItem(
+					$calendarEvent['objects'][0]['SUMMARY'][0] ?? 'New Event',
+					$this->dateTimeFormatter->formatTimeSpan(DateTime::createFromImmutable($startDate)),
+					$this->urlGenerator->getAbsoluteURL($this->urlGenerator->linkToRoute('calendar.view.index', ['objectId' => $calendarEvent['uid']])),
+					$this->urlGenerator->getAbsoluteURL($this->urlGenerator->linkToRoute('calendar.view.getCalendarDotSvg', ['color' => $calendar->getDisplayColor()])),
```

Comment:
`getDisplayColor` can return `null` and the route doesn't allow the parameter to be empty.